pmd-java/src/main/java/net/sourceforge/pmd/lang/java/qname/JavaTypeQualifiedName.java

```diff
@@ -0,0 +1,236 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.qname;
+
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Objects;
+
+import org.apache.commons.lang3.StringUtils;
+
+import net.sourceforge.pmd.lang.java.ast.JavaQualifiedName;
+import net.sourceforge.pmd.lang.java.typeresolution.PMDASMClassLoader;
+
+
+/**
+ * Specialises {@link JavaQualifiedName} for type names.
+ *
+ * @author Clément Fournier
+ * @since 6.1.0
+ */
+public final class JavaTypeQualifiedName extends JavaQualifiedName {
+
+    /** Local index value for when the class is not local. */
+    static final int NOTLOCAL_PLACEHOLDER = -1;
+
+    // Should we share that with ClassTypeResolver?
+    private static final PMDASMClassLoader CLASS_LOADER = PMDASMClassLoader.getInstance(JavaTypeQualifiedName.class.getClassLoader());
```

Comment:
Hm.... I think, we need to make sure, to use the auxclasspath here, if we use JavaTypeQualifiedName later for typeresolution...
In that case it can't be static final... we need to delegate it somehow from the QualifiedNameVisitor, similar how we do it for the type resolution phase (initializeWith...)